cxxfilt/__init__.py

```diff
@@ -1,117 +1,243 @@
+        >>> d
+        <Demangler libc='libc.so.6' libcxx='libstdc++.so.6'>
+        >>> d = cxxfilt.Demangler(find_library('c'), find_library('c++'))
+        >>> d
+        <Demangler libc='libc.so.6' libcxx='libc++.so.1'>
+    """
+
     def __init__(self, libc_name: str, libcxx_name: str) -> None:
+        """Creates a custom Demangler.
+
+        Args:
+            libc_name (str): The name/path to the libc library to be used.
+            libcxx_name (str): The name/path to the libcxx library to be used.
+
+        Raises:
+            AssertionError: If `libc_name` or `libcxx_name` is not a `str`
+        """
+
         assert isinstance(libc_name, str), libc_name
         assert isinstance(libcxx_name, str), libcxx_name
 
         self._libc_name = libc_name
         libc = ctypes.CDLL(libc_name)
-        self._free = libc.free
+        self._free = libc["free"]
         self._free.argtypes = [ctypes.c_void_p]
 
         self._libcxx_name = libcxx_name
         libcxx = ctypes.CDLL(libcxx_name)
-        # use getattr to workaround with python's own name mangling
-        self._cxa_demangle = getattr(libcxx, '__cxa_demangle')
-        self._cxa_demangle.restype = CharP
+        self._cxa_demangle = libcxx["__cxa_demangle"]
+        self._cxa_demangle.restype = ctypes.c_char_p
 
     def __repr__(self) -> str:
-        return f'<{self.__class__.__name__} libc={self._libc_name!r} libcxx={self._libcxx_name!r}>'
+        cls = self.__class__.__name__
+        libc = self._libc_name
+        libcxx = self._libcxx_name
+        return f"<{cls} libc={libc!r} libcxx={libcxx!r}>"
 
     def demangleb(self, mangled_name: bytes, external_only: bool = True) -> bytes:
         # Wikipedia: All *external* mangled symbols begin with _Z
-        if external_only and not mangled_name.startswith(b'_Z'):
+        if external_only and not mangled_name.startswith(b"_Z"):
             return mangled_name
 
         mangled_name_p = ctypes.c_char_p(mangled_name)
         status = ctypes.c_int()
-        retval = self._cxa_demangle(mangled_name_p, None, None, ctypes.pointer(status))
+        retval: ctypes.c_char_p = self._cxa_demangle(
+            mangled_name_p, None, None, ctypes.pointer(status)
+        )
 
         try:
             demangled = retval.value
         finally:
             self._free(retval)
 
-        if status.value == 0:
+        ret = status.value
+        if ret == 0:
             return demangled
-        elif status.value == -1:
-            raise InternalError('A memory allocation failiure occurred')
-        elif status.value == -2:
-            raise InvalidName(mangled_name)
-        elif status.value == -3:
-            raise InternalError('One of the arguments is invalid')
+        elif ret == -2:
+            raise InvalidNameError(mangled_name)
         else:
-            raise InternalError('Unkwon status code: {}'.format(status.value))
+            raise InternalError(ret)
```

Comment:
Do not move the status code handling into the `InternalError`. Having them all here makes it obvious how each return value is handled.
Also, the error codes aren't intended to be exposed to the user.